src/Hazelcast.Net/Serialization/ByteArrayObjectDataOutput.cs

```diff
@@ -363,13 +363,18 @@ public void Write(string value, Endianness endianness = Endianness.Unspecified)
         {
             endianness = endianness.Resolve(DefaultEndianness);
 
+            // length is the length of the string, in chars
+            // each char can be 1, 2 or 3 bytes - surrogate pairs are reported as 2 chars
+            // note: this is consistent with Java
+
             var length = value?.Length ?? ArraySerializer.NullArrayLength;
```

Comment:
Is this the UTF-8 length of the string or some sort of character count? For example, in Python `len("😀")` is equal to 1, but `len("😀".encode("utf-8"))` is equal to 4. We should be writing 4 as the length here because the encoded string will require this much space. Can you check this is writing the correct length?

Reply:
The length is the number of `char` items in the string, bearing in mind that every symbol that encodes on 1, 2 or 3 bytes is one `char` for C#, whereas symbols that encode on 4 bytes are two `char` (aka surrogate pair) - bit weird, but that is also what Java seems to do so AFAIK we would be compatible with Java here?

Reply:
I believe that could lead to some compatibility problems with other clients. I see that Java writes the length of the  UTF-8 encoded byte array (Python and Node.js also do that, not sure about C++) https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/ByteArrayObjectDataOutput.java#L261

Reply:
Then I mis-read the Java code, let me check